### ISSUE TITLE
fix or suppress build warnings on macos

### ISSUE THREAD

**Issue by semorrison** [leanprover/lean4#2541]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

Compiling anything using `lake` on macos systems reports warnings every time:

```
% lake init bar
% lake build
[0/3] Building Bar
[1/4] Compiling Bar
[1/4] Building Main
[2/4] Compiling Main
[4/4] Linking bar
stderr:
ld64.lld: warning: /usr/lib/system/libsystem_kernel.dylib has version 13.5.0, which is newer than target minimum of 13.0.0
ld64.lld: warning: /usr/lib/system/libsystem_platform.dylib has version 13.5.0, which is newer than target minimum of 13.0.0
ld64.lld: warning: /usr/lib/system/libsystem_pthread.dylib has version 13.5.0, which is newer than target minimum of 13.0.0
```

I had avoided reporting this one for a long time, as I had an old macos system and assumed it was my fault with mismatched OS / XCode / toolchain problems. But this still occurs (with different version numbers) on a fresh system.

### Additional Information

```
MACOSX_DEPLOYMENT_TARGET=13.5 lake build 
```
causes the problem to go away.

We need to work out a solution that does not require users to do this themselves, however. We can't have every new mac user quit Lean after `lake init bar && lake build` because they see error messages.

**Comment by tydeu** [comment]
This is a problem bundled compiler toolchain / `leanc` options. Windows also produces its own set of warnings with a `libgmp` installed:

```
[194/198] Compiling Lake.CLI
stderr:
cc: warning: C:/Strawberry/c/lib/libgmp.a: linker input file unused because linking not done
[194/198] Building Lake.Main
stderr:
cc: warning: C:/Strawberry/c/lib/libgmp.a: linker input file unused because linking not done
[197/198] Compiling Lake.Main
stderr:
cc: warning: C:/Strawberry/c/lib/libgmp.a: linker input file unused because linking not done
[198/198] Linking lake.exe
```

This error is printed every time a C file is compiled and the macOS is on every link.

**Comment by semorrison** [comment]
Can we set the flag automatically?

If not, can we filter the output in `lake`?

**Comment by tydeu** [comment]
@semorrison I think this just requires some fixes to the Lean build configuration. Thus, @Kha is probably the expert here.

**Comment by Kha** [comment]
This is presumably from https://github.com/leanprover/lean4/blob/63d2bdd4908b4df7db381537af754a7319e334e4/src/CMakeLists.txt#L452-L454. The intention was to make sure that Lean-compiled executables run on as many macOS versions as possible. But if it's not clear how to do that on an arbitrary user's machine, we should restrict it to building Lean itself and have users that care about it figure it out. To be honest, I don't know enough about macOS linking to tell whether it has any effect even for Lean itself.

**Comment by semorrison** [comment]
Unfortunately, @Kha, this has not resolved the issue. I am still seeing these warnings during `lake exe cache get` on Mathlib's `nightly-testing`, which uses your change from #2556.

**Comment by semorrison** [comment]
As before `MACOSX_DEPLOYMENT_TARGET=13.5 lake exe cache get` results in no errors.

**Comment by Kha** [comment]
Oh my, this looks like a clang+lld issue/bug: it knows how to derive the target version from the OS version, but only the major version: https://github.com/llvm/llvm-project/blob/a7b78cac9a77e3ef6bbbd8ab1a559891dc693401/llvm/lib/TargetParser/Triple.cpp#L1238C11-L1238C11
Possible stupid workaround: default to `MACOSX_DEPLOYMENT_TARGET=99.0`

**Comment by semorrison** [comment]
@Kha, I tried using `99.0` and it only makes things worse:
```
ld64.lld: warning: ~/.elan/toolchains/leanprover--lean4-pr-releases---pr-release-2593/lib/lean/libleanrt.a(object_ref.cpp.o) has version 99.0.0, which is newer than target minimum of 13.0.0
```
I don't really understand the message at all. It is warning that some number is *larger* than a specified *minimum*!

If we really can't get rid of these, we should suppress them in `lake`. It is not okay that compiling "hello world" prints error messages.

**Comment by Kha** [comment]
One last try: #2598

**Comment by semorrison** [comment]
Fixed in #2598
